internal/grpc: TestGRPCStreaming test flake

```
--- FAIL: TestGRPCStreaming (0.41s)
    entry.go:146: time="2018-09-25T13:52:11+10:00" level=info msg=stream_wait connection=1 error_detail=nil resource_names="[]" response_nonce= type_url=type.googleapis.com/envoy.api.v2.Cluster version_info=
    entry.go:146: time="2018-09-25T13:52:11+10:00" level=info msg=response connection=1 count=0 error_detail=nil resource_names="[]" response_nonce= type_url=type.googleapis.com/envoy.api.v2.Cluster version_info=
    entry.go:146: time="2018-09-25T13:52:11+10:00" level=info msg=stream_wait connection=1 error_detail=nil resource_names="[]" response_nonce= type_url=type.googleapis.com/envoy.api.v2.Cluster version_info=
    entry.go:146: time="2018-09-25T13:52:11+10:00" level=error msg="stream terminated" connection=1 error="context deadline exceeded"
    --- FAIL: TestGRPCStreaming/StreamClusters (0.10s)
        server_test.go:78: expected "DeadlineExceeded", got "Unknown"
    entry.go:146: time="2018-09-25T13:52:11+10:00" level=info msg=stream_wait connection=1 error_detail=nil resource_names="[]" response_nonce= type_url=type.googleapis.com/envoy.api.v2.ClusterLoadAssignment version_info=
    entry.go:146: time="2018-09-25T13:52:11+10:00" level=info msg=response connection=1 count=1 error_detail=nil resource_names="[]" response_nonce= type_url=type.googleapis.com/envoy.api.v2.ClusterLoadAssignment version_info=
    entry.go:146: time="2018-09-25T13:52:11+10:00" level=info msg=stream_wait connection=1 error_detail=nil resource_names="[]" response_nonce= type_url=type.googleapis.com/envoy.api.v2.ClusterLoadAssignment version_info=
    entry.go:146: time="2018-09-25T13:52:11+10:00" level=error msg="stream terminated" connection=1 error="context canceled"
    entry.go:146: time="2018-09-25T13:52:11+10:00" level=info msg=stream_wait connection=1 error_detail=nil resource_names="[]" response_nonce= type_url=type.googleapis.com/envoy.api.v2.Listener version_info=
    entry.go:146: time="2018-09-25T13:52:11+10:00" level=info msg=response connection=1 count=1 error_detail=nil resource_names="[]" response_nonce= type_url=type.googleapis.com/envoy.api.v2.Listener version_info=
    entry.go:146: time="2018-09-25T13:52:11+10:00" level=info msg=stream_wait connection=1 error_detail=nil resource_names="[]" response_nonce= type_url=type.googleapis.com/envoy.api.v2.Listener version_info=
    entry.go:146: time="2018-09-25T13:52:11+10:00" level=error msg="stream terminated" connection=1 error="context deadline exceeded"
    entry.go:146: time="2018-09-25T13:52:11+10:00" level=info msg=stream_wait connection=1 error_detail=nil resource_names="[]" response_nonce= type_url=type.googleapis.com/envoy.api.v2.RouteConfiguration version_info=
    entry.go:146: time="2018-09-25T13:52:11+10:00" level=info msg=response connection=1 count=2 error_detail=nil resource_names="[]" response_nonce= type_url=type.googleapis.com/envoy.api.v2.RouteConfiguration version_info=
    entry.go:146: time="2018-09-25T13:52:11+10:00" level=info msg=stream_wait connection=1 error_detail=nil resource_names="[]" response_nonce= type_url=type.googleapis.com/envoy.api.v2.RouteConfiguration version_info=
    entry.go:146: time="2018-09-25T13:52:11+10:00" level=error msg="stream terminated" connection=1 error="context deadline exceeded"
FAIL
FAIL    github.com/heptio/contour/internal/grpc 0.446s
```

The error we're getting is
```
--- FAIL: TestGRPC (0.42s)
    --- FAIL: TestGRPC/StreamClusters (0.10s)
        server_test.go:68: expected "DeadlineExceeded", got "Unknown" *status.statusError rpc error: code = Unknown desc = context deadline exceeded
```
the Code should be `DeadlineExceeded`, not `Unknown`.

Related, https://github.com/grpc/grpc-go/issues/1645

--- a/internal/grpc/server_test.go
+++ b/internal/grpc/server_test.go
@@ -95,9 +95,6 @@ func TestGRPC(t *testing.T) {
 			checktimeout(t, stream)          // check that the second receive times out
 		},
 		"StreamListeners": func(t *testing.T, cc *grpc.ClientConn) {
-			lds := v2.NewListenerDiscoveryServiceClient(cc)
-			ctx, cancel := context.WithTimeout(context.Background(), 100*time.Millisecond)
-			defer cancel()
 			// add an ingress, which will create a non tls listener
 			reh.OnAdd(&v1beta1.Ingress{
 				ObjectMeta: metav1.ObjectMeta{
@@ -120,6 +117,10 @@ func TestGRPC(t *testing.T) {
 					}},
 				},
 			})
+
+			lds := v2.NewListenerDiscoveryServiceClient(cc)
+			ctx, cancel := context.WithTimeout(context.Background(), 100*time.Millisecond)
+			defer cancel()
 			stream, err := lds.StreamListeners(ctx)
 			check(t, err)
 			sendreq(t, stream, listenerType) // send initial notification
@@ -275,7 +276,10 @@ func checktimeout(t *testing.T, stream interface {
 	if !ok {
 		t.Fatalf("%T %v", err, err)
 	}
-	if s.Code() != codes.DeadlineExceeded {
+
+	// Work around grpc/grpc-go#1645 which sometimes seems to
+	// set the status code to Unknown, even when the message is derived from context.DeadlineExceeded.
+	if s.Code() != codes.DeadlineExceeded && s.Message() != context.DeadlineExceeded.Error() {
 		t.Fatalf("expected %q, got %q %T %v", codes.DeadlineExceeded, s.Code(), err, err)
 	}
 }

